src/clp_ffi_js/ir/StreamReader.hpp

```diff
@@ -258,6 +295,39 @@ auto StreamReader::generic_filter_log_events(
         }
     }
 }
+
+template <typename LogEvent>
+requires requires(
+                 LogEventWithFilterData<LogEvent> const& event,
+                 clp::ir::epoch_time_ms_t timestamp
+         ) {
+    {
+        event.get_timestamp()
+    } -> std::convertible_to<clp::ir::epoch_time_ms_t>;
+}
+auto StreamReader::generic_get_log_event_index_by_timestamp(
+        LogEvents<LogEvent> const& log_events,
+        clp::ir::epoch_time_ms_t timestamp
+) -> LogEventIdxTsType {
+    if (log_events.empty()) {
+        return LogEventIdxTsType{emscripten::val::null()};
+    }
+    auto it = std::upper_bound(
+            log_events.begin(),
+            log_events.end(),
+            timestamp,
+            [](clp::ir::epoch_time_ms_t ts, LogEventWithFilterData<LogEvent> const& log_event) {
+                return ts < log_event.get_timestamp();
+            }
+    );
+
+    // If the iterator is not pointing to the beginning, decrement it so that it points to
+    // first element that is larger than timestamp.
+    if (it != log_events.begin()) {
+        --it;
+    }
+    return LogEventIdxTsType{emscripten::val(std::distance(log_events.begin(), it))};
```

Comment:
how about we rename the `it` as `upper` then this code becomes self-explanatory.

```suggestion
    if (upper == log_events.begin()) {
        return LogEventIdxTsType{0};
    }

    size_t upper_index{std::distance(log_events.begin(), upper)};
    const auto index{upper_index - 1};

    return LogEventIdxTsType{emscripten::val(index)};
```